### PR TITLE
Add zone info to instance name

### LINKED ISSUE #242: Explore possibility to use region details in OCP node names
Having PowerVS region identifier in OCP nodes will make it easier when trying to find out the OCP nodes in PowerVS UI

### PR DESCRIPTION
**Pull request by cs-zhang**
Fixed #242 

Add a flag `use_zone_info_for_instance`, the default is set to true.

Signed-off-by: CS Zhang <zhangcho@us.ibm.com>

### CODE REVIEW

**Comment by ltccci** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/ocp-power-automation/ocp4-upi-powervs/pull/271#" title="Author self-approved">cs-zhang</a>*
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **yussufsh** after the PR has been reviewed.
You can assign the PR to them by writing `/assign @yussufsh` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=ocp-power-automation%2Focp4-upi-powervs).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/ocp-power-automation/ocp4-upi-powervs/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: ocp.tf
Comment (bpradipt) [inline ocp.tf:9]: can you pls also fix the type: lenght ..

File: modules/1_prepare/variables.tf
Comment (bpradipt) [inline modules/1_prepare/variables.tf:27]: @cs-zhang overall looks good. Can you please update the docs as well to indicate the change

Comment (Prajyot-Parab) [reply modules/1_prepare/variables.tf:27]: @cs-zhang I am not clear about the length aspect, can you please share the reference doc where this is mentioned (length <= 24)

File: ocp.tf
Comment (cs-zhang) [reply ocp.tf:9]: Done.

File: modules/1_prepare/variables.tf
Comment (cs-zhang) [reply modules/1_prepare/variables.tf:27]: There is no doc for the name length limitation, but the name length for pi_network_name is limited to 32, its name created by `cluster_id` plus `-pub-net`, if over this limit, the TF will throw "Error: Failed to Create PI Network".

File: ocp.tf
Comment (cs-zhang) [reply ocp.tf:10]: The cluster_id_prefix null check is added.

**Review by bpradipt — DISMISSED** [review]
/lgtm

**Comment by ltccci** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/ocp-power-automation/ocp4-upi-powervs/pull/271#pullrequestreview-655912860" title="LGTM">bpradipt</a>*, *<a href="https://github.com/ocp-power-automation/ocp4-upi-powervs/pull/271#" title="Author self-approved">cs-zhang</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=ocp-power-automation%2Focp4-upi-powervs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/ocp-power-automation/ocp4-upi-powervs/blob/master/OWNERS)~~ [bpradipt]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by ltccci** [comment]
@bpradipt: The `/test` command needs one or more targets.
The following commands are available to trigger jobs:
* `/test pull-ocp4-upi-powervs-terraform-validate`
* `/test pull-ocp4-upi-powervs-terraform-fmt`
* `/test pull-ocp4-upi-powervs-terraform-lint`

Use `/test all` to run all jobs.


<details>

In response to [this](https://github.com/ocp-power-automation/ocp4-upi-powervs/pull/271#issuecomment-837042402):

>/test


Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

**Comment by ltccci** [comment]
@cs-zhang: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
pull-ocp4-upi-powervs-terraform-fmt | 5cb36de520b27de7a541c2ba46d3e02dc3acf434 | [link](https://prow.ppc64le-cloud.org/view/s3/prow-logs/pr-logs/pull/ocp-power-automation_ocp4-upi-powervs/271/pull-ocp4-upi-powervs-terraform-fmt/1392127018732097536) | `/test pull-ocp4-upi-powervs-terraform-fmt`

[Full PR test history](https://prow.ppc64le-cloud.org/pr-history?org=ocp-power-automation&repo=ocp4-upi-powervs&pr=271). [Your PR dashboard](https://prow.ppc64le-cloud.org/pr?query=is:pr+state:open+author:cs-zhang).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by mkumatag** [comment]
> @cs-zhang: The following test **failed**, say `/retest` to rerun all failed tests:
> 
> Test name	Commit	Details	Rerun command
> pull-ocp4-upi-powervs-terraform-fmt	[5cb36de](https://github.com/ocp-power-automation/ocp4-upi-powervs/commit/5cb36de520b27de7a541c2ba46d3e02dc3acf434)	[link](https://prow.ppc64le-cloud.org/view/s3/prow-logs/pr-logs/pull/ocp-power-automation_ocp4-upi-powervs/271/pull-ocp4-upi-powervs-terraform-fmt/1392127018732097536)	`/test pull-ocp4-upi-powervs-terraform-fmt`
> [Full PR test history](https://prow.ppc64le-cloud.org/pr-history?org=ocp-power-automation&repo=ocp4-upi-powervs&pr=271). [Your PR dashboard](https://prow.ppc64le-cloud.org/pr?query=is:pr+state:open+author:cs-zhang).

There is fmt error, need to be fixed:

image

**Review by bpradipt — DISMISSED** [review]
/lgtm

**Comment by ltccci** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/ocp-power-automation/ocp4-upi-powervs/pull/271#pullrequestreview-657050440" title="LGTM">bpradipt</a>*, *<a href="https://github.com/ocp-power-automation/ocp4-upi-powervs/pull/271#" title="Author self-approved">cs-zhang</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=ocp-power-automation%2Focp4-upi-powervs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/ocp-power-automation/ocp4-upi-powervs/blob/master/OWNERS)~~ [bpradipt]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: ocp.tf
Comment (yussufsh) [reply ocp.tf]: @cs-zhang  I really feel user should have ability to not use the zone in the cluster name. User should be able to not use it some how?

Comment (bpradipt) [reply ocp.tf]: @yussufsh typically in cloud instances the zone name helps to find out where the nodes exist. So making it default with cluster-id will help here imho.
Are you suggesting to make adding of zone in the prefix as optional ?

Comment (yussufsh) [reply ocp.tf]: Yes some customer may not want powervs zone names in their cluster URIs. A flag to include zone in cluster id can help which could be true by default.

Comment (bpradipt) [reply ocp.tf]: hmm, this makes me wonder if there is a way to use the zone  for the host names by default and make it optional for the cluster URL ?
@cs-zhang any thoughts ?

File: modules/1_prepare/variables.tf
Comment (cs-zhang) [reply modules/1_prepare/variables.tf:27]: Keep `-pub-net ` is to show these networks are created by TF automation.  It can be removed if we don't care to distinguish them from other networks.

Comment (gerrith3) [reply modules/1_prepare/variables.tf:27]: I don't see the value of keeping -pub-net - isn't everything created by TF automation?  There may be other networks in the service area but they are probably named for human readability and would be hard to confuse with an automatically created network?  I'd prefer as many characters as possible for these names, and 32 is much better than 24 or 14.

**Comment by bpradipt** [comment]
@cs-zhang you'll need to rebase

**Comment by ltccci** [comment]
New changes are detected. LGTM label has been removed.

**Comment by ltccci** [comment]
@cs-zhang: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
pull-ocp4-upi-powervs-terraform-fmt | 3096e7453cbfbc8ff8fa53e9bb8aec94d92b26bc | [link](https://prow.ppc64le-cloud.org/view/s3/prow-logs/pr-logs/pull/ocp-power-automation_ocp4-upi-powervs/271/pull-ocp4-upi-powervs-terraform-fmt/1394372286080880640) | `/test pull-ocp4-upi-powervs-terraform-fmt`

[Full PR test history](https://prow.ppc64le-cloud.org/pr-history?org=ocp-power-automation&repo=ocp4-upi-powervs&pr=271). [Your PR dashboard](https://prow.ppc64le-cloud.org/pr?query=is:pr+state:open+author:cs-zhang).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Review by bpradipt — DISMISSED** [review]
/lgtm
Thanks @cs-zhang

**Comment by ltccci** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/ocp-power-automation/ocp4-upi-powervs/pull/271#pullrequestreview-662835013" title="LGTM">bpradipt</a>*, *<a href="https://github.com/ocp-power-automation/ocp4-upi-powervs/pull/271#" title="Author self-approved">cs-zhang</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=ocp-power-automation%2Focp4-upi-powervs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/ocp-power-automation/ocp4-upi-powervs/blob/master/OWNERS)~~ [bpradipt]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: modules/5_install/install.tf
Comment (yussufsh) [inline modules/5_install/install.tf:87]: Removing domain will not work for external dns.

**Review by Prajyot-Parab — COMMENTED** [review]
/lgtm

**Comment by Prajyot-Parab** [comment]
CC- @yussufsh

File: modules/1_prepare/prepare.tf
Comment (yussufsh) [inline modules/1_prepare/prepare.tf:140]: Do we need to use the prefix in hostname as well? I do not think so. let the hostname be with cluster id, just the instance name can have zone.

File: ocp.tf
Comment (yussufsh) [inline ocp.tf:18]: 1. Rename it to `name_suffix`
2. Just pass the suffix part and use `"${var.cluster_id}${var.name_suffix}"` everywhere.
```
name_suffix = var.use_zone_info_for_instance ? "-${var.ibmcloud_zone}" : ""
```

This will ensure the existing flow for cluster_id is not impacted. For example no_proxy, bastion name for external DNS,. etc.

File: modules/7_ibmcloud/dns.tf
Comment (yussufsh) [inline modules/7_ibmcloud/dns.tf:33]: This is not needed if we are not changing the hostname and sticking with only resource name change for adding zone.

File: modules/7_ibmcloud/variables.tf
Comment (yussufsh) [inline modules/7_ibmcloud/variables.tf:27]: Do we want to use the zone in load_balancer name?
I understand this is VPC but to identify the cluster using this LB is in which zone may help.
@bpradipt

File: modules/5_install/install.tf
Comment (yussufsh) [inline modules/5_install/install.tf:44]: This is not needed if we are not changing the hostname and sticking with only resource name change for adding zone.

File: docs/var.tfvars-doc.md
Comment (yussufsh) [inline docs/var.tfvars-doc.md:187]: The variable name mentions only about instance. We are also using it for network, keypair, volume name as well.
Suggestion for name: use_zone_for_names

**Review by yussufsh — CHANGES_REQUESTED** [review]
This review is based on the assumption that we are now only using the zone in the resource naming on PowerVS. We need to make sure the cluster_id is not impacted and the cluster naming is same as before with no change in cluster details.

If we are changing the hostnames then there are many other places you will need to change, that will also change the cluster details. Hence it is a good idea to stick with resource naming with zone and not use it in the cluster.

File: ocp.tf
Comment (yussufsh) [reply ocp.tf:18]: I am even okay with `cluster_id_with_zone` if you are planning to include `local.cluster_id` to it. This will also help in validating the length of 32 with zone info.

File: modules/1_prepare/prepare.tf
Comment (bpradipt) [reply modules/1_prepare/prepare.tf:140]: having hostname same as instance name will avoid confusion imho

File: modules/1_prepare/variables.tf
Comment (yussufsh) [inline modules/1_prepare/variables.tf:33]: Seems you have increased the cluster_id length to 32 without considering `name_prefix` that has the zone info. We will need to see if `name_prefix` (cluster_id+zone) is <= 32

File: modules/1_prepare/prepare.tf
Comment (yussufsh) [reply modules/1_prepare/prepare.tf:140]: This will need changes in code for helpernode and install playbook vars, also in the external dns as well. Also imo it will add to more confusion with cluster name having the zone. So my suggestion is to stick with zone only in resource naming and not impact the cluster naming.

If we want to use zone everywhere, ie: names, hostnames, cluster name, URIs, then no need to add new variable called name_prefix. Append the zone to cluster_id with no changes elsewhere.

Comment (yussufsh) [reply modules/1_prepare/prepare.tf:140]: Closing this comment as we decided to have zone name in hostname for the nodes.

File: modules/7_ibmcloud/dns.tf
Comment (yussufsh) [reply modules/7_ibmcloud/dns.tf:33]: Closing this comment as we decided to have zone name in hostname for the nodes.

File: modules/5_install/install.tf
Comment (yussufsh) [reply modules/5_install/install.tf:44]: Closing this comment as we decided to have zone name in hostname for the nodes.

File: modules/7_ibmcloud/variables.tf
Comment (bpradipt) [reply modules/7_ibmcloud/variables.tf:27]: @yussufsh ok

File: docs/var.tfvars-doc.md
Comment (cs-zhang) [reply docs/var.tfvars-doc.md:187]: Done.

File: modules/7_ibmcloud/variables.tf
Comment (cs-zhang) [reply modules/7_ibmcloud/variables.tf:27]: zone info added for LB and security group.

File: modules/5_install/install.tf
Comment (cs-zhang) [reply modules/5_install/install.tf:87]: If the installation failed with domain.  Why just add the domain without cluster_id for external dns? The bastion hostname should be xxxx-bastion-x.`cluster_id`.`cluster_domain`, rIght?

File: docs/var.tfvars-doc.md
Comment (bpradipt) [reply docs/var.tfvars-doc.md:187]: @cs-zhang don't see the latest changes.. Did you push it ?

**Comment by ltccci** [comment]
New changes are detected. LGTM label has been removed.

**Comment by ltccci** [comment]
@cs-zhang: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
pull-ocp4-upi-powervs-terraform-validate | a3f78333cded71f83ec403a8fde5148ba2232324 | [link](https://prow.ppc64le-cloud.org/view/s3/prow-logs/pr-logs/pull/ocp-power-automation_ocp4-upi-powervs/271/pull-ocp4-upi-powervs-terraform-validate/1395141323094233088) | `/test pull-ocp4-upi-powervs-terraform-validate`

[Full PR test history](https://prow.ppc64le-cloud.org/pr-history?org=ocp-power-automation&repo=ocp4-upi-powervs&pr=271). [Your PR dashboard](https://prow.ppc64le-cloud.org/pr?query=is:pr+state:open+author:cs-zhang).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

File: modules/1_prepare/variables.tf
Comment (yussufsh) [inline modules/1_prepare/variables.tf:36]: This should be `<= 32`

**Review by yussufsh — CHANGES_REQUESTED** [review]

File: docs/var.tfvars-doc.md
Comment (cs-zhang) [reply docs/var.tfvars-doc.md:196]: Here it should not include `ibmcloud_zone`, and it is for` cloud_id` only.

File: modules/1_prepare/variables.tf
Comment (cs-zhang) [reply modules/1_prepare/variables.tf]: done

File: modules/1_prepare/variables.tf
Comment (cs-zhang) [reply modules/1_prepare/variables.tf]: Done.

File: docs/var.tfvars-doc.md
Comment (yussufsh) [reply docs/var.tfvars-doc.md:196]: But the name prefix is made up of all 3 values right? It is not known to user that you are automatically adding zone info to the name. So name=cluster_id_prefix+cluster_id+zone, where the length should not exceed 32. So here cluster_id_prefix+cluster_id cannot exceed 32-length(zone).

Comment (cs-zhang) [reply docs/var.tfvars-doc.md:196]: I have added that to below for `use_zone_info_for_name`.

**Comment by yussufsh** [comment]
@cs-zhang please squash you commits into one, I will approve.

**Review by bpradipt — APPROVED** [review]
/lgtm

@yussufsh please recheck

**Comment by ltccci** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/ocp-power-automation/ocp4-upi-powervs/pull/271#pullrequestreview-669740751" title="LGTM">bpradipt</a>*, *<a href="https://github.com/ocp-power-automation/ocp4-upi-powervs/pull/271#" title="Author self-approved">cs-zhang</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=ocp-power-automation%2Focp4-upi-powervs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/ocp-power-automation/ocp4-upi-powervs/blob/master/OWNERS)~~ [bpradipt]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Review by yussufsh — APPROVED** [review]
/lgtm
cc: @sudeeshjohn FYI this will add new variable to prepare module once back-ported to release branches.

**Comment by ltccci** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/ocp-power-automation/ocp4-upi-powervs/pull/271#pullrequestreview-669740751" title="LGTM">bpradipt</a>*, *<a href="https://github.com/ocp-power-automation/ocp4-upi-powervs/pull/271#" title="Author self-approved">cs-zhang</a>*, *<a href="https://github.com/ocp-power-automation/ocp4-upi-powervs/pull/271#pullrequestreview-669764920" title="LGTM">yussufsh</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=ocp-power-automation%2Focp4-upi-powervs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/ocp-power-automation/ocp4-upi-powervs/blob/master/OWNERS)~~ [bpradipt,yussufsh]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Pull request merged**
